core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java

```diff
@@ -261,27 +259,11 @@ private static String decodeUtf8Component(TemporaryThreadLocals tempThreadLocals
     }
 
     private static int decodeHexByte(char c1, char c2) {
-        final int hi = decodeHexNibble(c1);
-        final int lo = decodeHexNibble(c2);
+        final int hi = OCTETS_TO_HEX[c1];
```

Comment:
Don't you still want to use PlatformDependent throughout this PR to avoid automatic bounds check too? Maybe we can have a method that accepts char, `assert` array length is less than 64K, and then do the unsafe lookup

Reply:
JVM already eliminates the automatic bounds check because it knows c1 and c2 is always < 65536 and `OCTETS_TO_HEX.length` is 65536.

Reply:
Checked printed assembly and confirmed both indexing and getByte produce the same code!

Reply:
Sorry I looked again - I think I see a bounds check at `0x000001ab07adaa2c:   cmp    $0x10000,%edx`. Unsafe version doesn't have cmp or any fallback codepath.

Array index

```
============================= C2-compiled nmethod ==============================
<nmethod compile_id='39' compiler='c2' entry='0x000002e5c6c28aa0' size='664' address='0x000002e5c6c28910' relocation_offset='360' insts_offset='400' stub_offset='496' scopes_data_offset='544' scopes_pcs_offset='576' dependencies_offset='656' oops_offset='520' metadata_offset='536' method='com.linecorp.armeria.jvm.ArraysBenchmark arrayIndex (C)B' bytes='6' count='5179' backedge_count='1' iicount='100161' stamp='0.340'/>
<print_nmethod compile_id='39' compiler='c2' stamp='0.340'>
----------------------------------- Assembly -----------------------------------

Compiled method (c2)     340   39             com.linecorp.armeria.jvm.ArraysBenchmark::arrayIndex (6 bytes)
 total in heap  [0x000002e5c6c28910,0x000002e5c6c28ba8] = 664
 relocation     [0x000002e5c6c28a78,0x000002e5c6c28a88] = 16
 main code      [0x000002e5c6c28aa0,0x000002e5c6c28b00] = 96
 stub code      [0x000002e5c6c28b00,0x000002e5c6c28b18] = 24
 oops           [0x000002e5c6c28b18,0x000002e5c6c28b28] = 16
 metadata       [0x000002e5c6c28b28,0x000002e5c6c28b30] = 8
 scopes data    [0x000002e5c6c28b30,0x000002e5c6c28b50] = 32
 scopes pcs     [0x000002e5c6c28b50,0x000002e5c6c28ba0] = 80
 dependencies   [0x000002e5c6c28ba0,0x000002e5c6c28ba8] = 8

--------------------------------------------------------------------------------
[Constant Pool (empty)]

--------------------------------------------------------------------------------

[Verified Entry Point]
  # {method} {0x000002e5ce762728} &apos;arrayIndex&apos; &apos;(C)B&apos; in &apos;com/linecorp/armeria/jvm/ArraysBenchmark&apos;
  # parm0:    rdx       = char
  #           [sp+0x20]  (sp of caller)
  0x000002e5c6c28aa0:   mov    %eax,-0x7000(%rsp)
  0x000002e5c6c28aa7:   push   %rbp
  0x000002e5c6c28aa8:   sub    $0x10,%rsp                   ;*synchronization entry
                                                            ; - com.linecorp.armeria.jvm.ArraysBenchmark::arrayIndex@-1 (line 55)
  0x000002e5c6c28aac:   cmp    $0x10000,%edx
  0x000002e5c6c28ab2:   jae    0x000002e5c6c28ad4
  0x000002e5c6c28ab4:   movabs $0xefc18840,%r10             ;   {oop([B{0x00000000efc18840})}
  0x000002e5c6c28abe:   movsbl 0x10(%r10,%rdx,1),%eax
  0x000002e5c6c28ac4:   add    $0x10,%rsp
  0x000002e5c6c28ac8:   pop    %rbp
  0x000002e5c6c28ac9:   mov    0x108(%r15),%r10
  0x000002e5c6c28ad0:   test   %eax,(%r10)                  ;   {poll_return}
  0x000002e5c6c28ad3:   retq
  0x000002e5c6c28ad4:   mov    %edx,%ebp
  0x000002e5c6c28ad6:   mov    $0xffffffe4,%edx
  0x000002e5c6c28adb:   callq  0x000002e5c6be9080           ; ImmutableOopMap {}
                                                            ;*baload {reexecute=0 rethrow=0 return_oop=0}
                                                            ; - com.linecorp.armeria.jvm.ArraysBenchmark::arrayIndex@4 (line 55)
                                                            ;   {runtime_call UncommonTrapBlob}
  0x000002e5c6c28ae0:   ud2                                 ;*baload {reexecute=0 rethrow=0 return_oop=0}
                                                            ; - com.linecorp.armeria.jvm.ArraysBenchmark::arrayIndex@4 (line 55)
  0x000002e5c6c28ae2:   hlt

```

Unsafe version

```
============================= C2-compiled nmethod ==============================
<nmethod compile_id='233' compiler='c2' entry='0x0000019c552aafa0' size='632' address='0x0000019c552aae10' relocation_offset='360' insts_offset='400' stub_offset='464' scopes_data_offset='528' scopes_pcs_offset='560' dependencies_offset='624' oops_offset='488' metadata_offset='496' method='com.linecorp.armeria.jvm.ArraysBenchmark unsafeArrayLookup (C)B' bytes='8' count='5400' backedge_count='1' iicount='60390' stamp='2.793'/>
<print_nmethod compile_id='233' compiler='c2' stamp='2.793'>
----------------------------------- Assembly -----------------------------------

Compiled method (c2)    2792  233             com.linecorp.armeria.jvm.ArraysBenchmark::unsafeArrayLookup (8 bytes)
 total in heap  [0x0000019c552aae10,0x0000019c552ab088] = 632
 relocation     [0x0000019c552aaf78,0x0000019c552aaf88] = 16
 main code      [0x0000019c552aafa0,0x0000019c552aafe0] = 64
 stub code      [0x0000019c552aafe0,0x0000019c552aaff8] = 24
 oops           [0x0000019c552aaff8,0x0000019c552ab000] = 8
 metadata       [0x0000019c552ab000,0x0000019c552ab020] = 32
 scopes data    [0x0000019c552ab020,0x0000019c552ab040] = 32
 scopes pcs     [0x0000019c552ab040,0x0000019c552ab080] = 64
 dependencies   [0x0000019c552ab080,0x0000019c552ab088] = 8

--------------------------------------------------------------------------------
[Constant Pool (empty)]

--------------------------------------------------------------------------------

[Verified Entry Point]
  # {method} {0x0000019c5c5428a8} &apos;unsafeArrayLookup&apos; &apos;(C)B&apos; in &apos;com/linecorp/armeria/jvm/ArraysBenchmark&apos;
  # parm0:    rdx       = char
  #           [sp+0x20]  (sp of caller)
  0x0000019c552aafa0:   sub    $0x18,%rsp
  0x0000019c552aafa7:   mov    %rbp,0x10(%rsp)              ;*synchronization entry
                                                            ; - com.linecorp.armeria.jvm.ArraysBenchmark::unsafeArrayLookup@-1 (line 64)
  0x0000019c552aafac:   movslq %edx,%r10
  0x0000019c552aafaf:   movabs $0xee300000,%r11             ;   {oop([B{0x00000000ee300000})}
  0x0000019c552aafb9:   movsbl 0x10(%r11,%r10,1),%eax       ;*invokevirtual getByte {reexecute=0 rethrow=0 return_oop=0}
                                                            ; - sun.misc.Unsafe::getByte@5 (line 230)
                                                            ; - io.netty.util.internal.PlatformDependent0::getByte@10 (line 525)
                                                            ; - io.netty.util.internal.PlatformDependent::getByte@2 (line 503)
                                                            ; - com.linecorp.armeria.jvm.ArraysBenchmark::unsafeArrayLookup@4 (line 64)
  0x0000019c552aafbf:   add    $0x10,%rsp
  0x0000019c552aafc3:   pop    %rbp
  0x0000019c552aafc4:   mov    0x108(%r15),%r10
  0x0000019c552aafcb:   test   %eax,(%r10)                  ;   {poll_return}
  0x0000019c552aafce:   retq
  0x0000019c552aafcf:   hlt
  0x0000019c552aafd0:   hlt
  0x0000019c552aafd1:   hlt
  0x0000019c552aafd2:   hlt
```

Code
```
package com.linecorp.armeria.jvm;

import java.util.Arrays;
import java.util.Random;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;

import io.netty.util.internal.PlatformDependent;

/**
 * Benchmarks to check JVM optimizations of using arrays.
 */
@State(Scope.Benchmark)
public class ArraysBenchmark {

    private static final byte[] LARGER_THAN_SIXTEEN_BITS = new byte[Character.MAX_VALUE + 1];

    static {
        Arrays.fill(LARGER_THAN_SIXTEEN_BITS, (byte) -1);
    }

    private char index;

    @Setup
    public void setUp() {
        // Use fixed seed random to create an int, the JVM should have no way of constant-folding.
        index = (char) new Random().nextInt(Character.MAX_VALUE + 1);
    }

    @Benchmark
    public byte arrayIndex() {
        return arrayIndex(index);
    }

    private static byte arrayIndex(char index) {
        return LARGER_THAN_SIXTEEN_BITS[index];
    }

    @Benchmark
    public byte unsafeArrayLookup() {
        return unsafeArrayLookup(index);
    }

    private static byte unsafeArrayLookup(char index) {
        return PlatformDependent.getByte(LARGER_THAN_SIXTEEN_BITS, index);
    }

    @SuppressWarnings("checkstyle:UncommentedMain")
    public static void main(String[] args) {
        final ArraysBenchmark benchmark = new ArraysBenchmark();
        byte res = 0;
        for (int i = 0; i < 100_000_000; i++) {
            benchmark.setUp();
            res = benchmark.arrayIndex();
        }
        System.out.println(res);
    }
}
```

Reply:
Why not using `-prof perfasm` directly on JMH? It should be the same eh...just curious

Reply:
I see this too:
```
// Use fixed seed random to create an int, the JVM should have no way of constant-folding.
```
It isn't necessary because the JVM trust just `final static` fields to perform constant folding (or `final` on the stack, or inside special classes/captured effective-final on lambdas...)

Reply:
I'm currently on Windows so it'd take some extra effort to set up perf, at least I think so 😆

Reply:
Let me take a look on `DecodeHexBenchmark.hexDigitsSafeNoChecks` and `DecodeHexBenchmark.hexDigitsUnsafeNoChecks` ASM (on https://github.com/netty/netty/pull/9896)  what's going on

Reply:
You're probably right about the folding, I just wanted to make extra-sure for now since just looking at the assembly and not checking benchmark numbers yet.

FWIW, if I'm not mistaken, when indexing from a `char` _field_, there isn't a bounds check but when indexing from a `char` method _parameter_, there is. I initially thought the code was the same before splitting out a method with `char` param to better match the code, and then I got the above. It does seem to "know" the parameter is char (`# parm0:    rdx       = char`), but still has a jump.

Reply:
By the way, I'm literally searching through `hotspot_pid*.log` in less to check the assembly as my jitwatch also tells me to enable class loading trace, even though it is enabled... So I could be misreading something for sure. But one takeaway may be that as JVM optimizations are a black box, if we have a more direct method `PlatformDependent.getByte` it probably doesn't hurt to just use it :) Actually, if `PlatformDependent.getByte` did the `hasUnsafe` check the code doesn't really get cluttered either, would that be a reasonable change (`static final boolean` should be removed completely even by C1)? Actually I'm not sure what's platform dependent about that method ;)

Reply:
That's strange: you can try verify this by not making `arrayIndex` to be inlined using `@CompilerControl(Mode.DONT_INLINE)` on it

Reply:
I think that Android miss `Unsafe` in the form we use it for other JVMs: my concern is that unsafe manipulation of arrays (or any field in an proper Java Object) *could* prevent certain nice optimizations related vectorization due to the introduction of compiler barriers: I don't remember the JEP and if has been solved, but I remember wasn't something 100% safe to be assumed. Anyway the numbers are not so different and is yet to be seen with the core cache filled of other stuff, in a real case, how much this optimization will make difference